src/providers/hcloud/mod.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! aws ec2 metadata fetcher
+//!
+
+use std::collections::HashMap;
+
+use openssh_keys::PublicKey;
+use update_ssh_keys::AuthorizedKeyEntry;
+
+use errors::*;
+use network;
+use providers::MetadataProvider;
+use retry;
+use serde_json;
+
+#[cfg(not(test))]
+const URL: &str = "http://169.254.169.254/2009-04-04";
```

Comment:
Is there a public reference where this endpoint JSON response is documented?

Reply:
@LKaemmerling Could you answer this from the Hetzner team? It seems there is indeed no public documentation of this endpoint (i.e. on https://docs.hetzner.cloud/)

Reply:
This endpoint is not public documented.

Reply:
Since it is not public documented, we can not guarantee that it will behave the same way as the current implementation.

Reply:
Thanks for your answer and checking with your colleagues. It would indeed be good to have public documentation somewhere. Currently the only information used in the coreos-metadata are hostname, public_keys and public ipv4 address. 

These can come either from http://169.254.169.254/2009-04-04/meta-data/hostname etc or from the YAML from http://169.254.169.254/hetzner/v1/metadata. 

Could you maybe ask the Datacenter Team if they have an idea whether or not the breaking changes in future they are thinking of would affect any of these three things or not. If not, there is no problem here. If they do, I believe it's better to either wait for your metaservice to be stable and publicly documented - or change the coreos-metadata implementation at the moment the change is released? (or in anticipation of it)

Reply:
I think it would be better to wait for the stable and documented metadata service, but I can not say you an ETA.

Reply:
@LKaemmerling One final thought... 

The current endpoints that are used are:
- http://169.254.169.254/2009-04-04/meta-data/hostname
- http://169.254.169.254/2009-04-04/meta-data/public-ipv4
- http://169.254.169.254/2009-04-04/meta-data/local-ipv4
- http://169.254.169.254/2009-04-04/meta-data/public-keys
- http://169.254.169.254/2009-04-04/meta-data/instance-id

So, all under the "2009-04-04" version of the service. What usually happens when a new version is released, you put this under a new url i.e. http://169.254.169.254/2018-11-29/ similar to the versioning done at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-meta-data-retrieval-examples Otherwise all services using scripts relying on this endpoint to be available i.e. to identify itself or identify on which host they are running, would break.

If you could confirm this could be the case, I see no reason to wait as the current version has all the information needed to allow CoreOS instances to smoothly boot in the Hetzner Cloud.

Reply:
It would be better to wait, because it can be possible that we drop the ec2-ish metadata. (This is only one option, not fixed!) I will write you when I have more information.

Reply:
@LKaemmerling I noticed the official CSI Driver for Kubernetes has been published. As the nodes needs to identify itself using a meta service, I can see you are using the http://169.254.169.254/2009-04-04/meta-data/instance-id endpoint: 

https://github.com/hetznercloud/csi-driver/blob/8113aa5f6026e4bd75ef225cae3c0c6eb6063601/src/cmd/driver/main.go#L165

As this is now also used within an official Hetzner supported driver, can't we use the same endpoint for this integration too?

Reply:
I see the official and stable Hetzner Cloud CSI driver for Kubernetes has been released (https://github.com/hetznercloud/csi-driver/commit/cf02070d4339b472560ffc6aadf9709fac4965df) which is using the same API endpoint the CoreOS Metadata integration would use (https://github.com/hetznercloud/csi-driver/blob/master/src/cmd/driver/main.go#L166). 

From his status, @LKaemmerling seems to be out for an unknown period (@LKaemmerling I hope you recover and get well soon).. Maybe @thcyron can comment and confirm the  http://169.254.169.254 may be used so we can move forward with these PRs?

Reply:
You can use and rely on these endpoints.

Reply:
@thcyron thanks for the feedback! Do you have some spec documentation URL to reference here, or should we just follow whatever the CSI driver is doing?

Reply:
Unfortunately our metadata service is not documented. You can use all endpoints except `local-ipv4` (since we don’t have private networking yet) as listed in [this comment](
https://github.com/coreos/coreos-metadata/pull/125#discussion_r237610165).

Reply:
Should we remove the `local-ipv4` or can we keep it for now? For the moment it is not used anywhere but then the metadata service in coreos is already prepared for the moment Hetzner would add private networking (which would be awesome ;))?

Reply:
I think you can keep it. It’s no secret that we’ll offer private networking sooner or later and I’m pretty sure we’ll support the `local-ipv4` endpoint then.